### PR TITLE
[Draft] initial pass on registry event API

### PR DESCRIPTION
**Pull request by asiekierka**
i hate IdList i hate IdList i hate IdList

### CODE REVIEW

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java [hunk 1]
```diff
@@ -14,11 +14,23 @@
...
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.impl.registry.ListenableRegistry;
 import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
 
 @FunctionalInterface
-public interface RegistryPostRegisterCallback<T> extends RegistryCallback<T> {
-	void onPostRegister(int rawId, Identifier id, T object);
+public interface RegistryAddObjectCallback<T> {
+	void onAddObject(int rawId, Identifier id, T object);
```
Comment (asiekierka) [inline fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: There should be a matching RegistryReplaceObjectCallback

Comment (liach) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: We should not allow registry replacement. This was promoted by forge, but in fabric, we should ask modders to inject onto the code of the object they are modifying. That way, if two mods add hooks to leaves at the same time (changing two different properties, for example), they can be both applied, unlike in forge that only one is preserved.

Comment (JamiesWhiteShirt) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: We don't always have a choice to efficiently inject into the code when a block or item is an instance of a shared class. Both options have their use cases.

Comment (modmuss50) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: There are good reasons to replace, and there are some good reasons not to. I don’t see any reason I’m not allowing it, it’s up to the mod to try and be as compatible as possible.

Comment (liach) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: Sure. A warning like that in forge imo is the best.

Comment (asiekierka) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: I'm talking about the *callback*.

Though, to minimize the amount of work on mods' side, we should probably just emit Remove, then Add.

Comment (asiekierka) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: Handling such a situation does not imply supporting such a situation - if another loader below performs a replacement, you'd still want to know, no?

Comment (asiekierka) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryAddObjectCallback.java]: Done.

**Comment by asiekierka** [comment]
So, I'm generally pleased with the shape of this (helper methods can come later) - I am more curious if everyone's fine with the way the remapping callback works.

**Comment by InsomniaKitten** [comment]
I think `Entry` would be better than `Object` for this

**Comment by asiekierka** [comment]
Done. As I said, looking for thoughts on the callbacks as they are provided now.

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryRemapCallback.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import net.fabricmc.fabric.impl.registry.ListenableRegistry;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+/**
+ * The remapping process functions as follows:
+ *
+ * - RegistryRemoveObjectCallbacks are called to remove any objects culled in the process, with the old numeric ID.
+ * - RegistryRemapCallback is emitted to allow remapping the IDs of objects still present.
+ * - RegistryAddObjectCallbacks are called to add any objects added in the process, with the new numeric ID.
```
Comment (Boundarybreaker) [inline fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryRemapCallback.java]: RegistryRemoveEntryCallback/RegistryAddEntryCallback, right?

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java [hunk 3]
```diff
@@ -235,29 +248,40 @@ public void remap(String name, Object2IntMap<Identifier> remoteIndexedEntries, R
...
 
 	@Override
 	public void unmap(String name) throws RemapException {
 		if (fabric_prevIndexedEntries != null) {
+			List<Identifier> addedIds = new ArrayList<>();
+
+			// Emit AddObject events for previously culled objects.
+			for (Identifier id : fabric_prevEntries.keySet()) {
+				if (!entries.containsKey(id)) {
+					assert fabric_prevIndexedEntries.containsKey(id);
```
Comment (Boundarybreaker) [inline fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/MixinIdRegistry.java]: Would this be better thrown as something other than an AssertionException?

File: fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryRemapCallback.java (hunk 2 above)
Comment (asiekierka) [reply fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/RegistryRemapCallback.java]: Yes, I missed the rename

**Comment by liach** [comment]
:tada:

**Pull request merged**